api/src/wfl/source.clj

```diff
@@ -240,20 +241,32 @@
         {:status status
          :body   (util/response-body-json data)}))))
 
-(defn ^:private check-tdr-job
+(defn ^:private tdr-job-failed-slack-msg
+  "Return a mrkdwn Slack message to be emitted for failed TDR jobs."
+  [{:keys [id job_status] :as _metadata}
+   {:keys [status body]   :as _caught-job-result}]
+  (str/join \newline
+            [(format "*:sadpanda: Snapshot creation job %s %s*" id job_status)
+             (format "`%s: %s`" status (:message body))]))
+
+(defn ^:private check-tdr-job-and-notify-on-failure
   "Check TDR job status for `job-id` and return job metadata,
-   with snapshot_id attached if the job succeeded."
+   with snapshot_id attached if the job succeeded.
+   Notify `workload` watchers if the job failed."
   [workload job-id]
   (let [{:keys [job_status] :as metadata} (datarepo/job-metadata job-id)
         get-job-result                    #(datarepo/job-result job-id)]
     (case job_status
       "running"   metadata
       "succeeded" (assoc metadata :snapshot_id (:id (get-job-result)))
-      (do (log/warning "TDR job failed or has otherwise unknown status"
-                       :workload   (workloads/to-log workload)
-                       :metadata    metadata
-                       :job-result (result-or-catch get-job-result))
-          metadata))))
+      (let [caught  (result-or-catch get-job-result)
```

Comment:
I retained your advice: "If using a `(do …)` consider if you can use a `(let …)` :)

It did clue me into how this block is pretty much side effects soup though.

Reply:
OK. Thanks.
Note swapping the arguments to `[job_id workload]` simplifies calling this function.